Split up particles over multiple tiles for OpenMP

atmyers opened the pull request:
I now get, on [this](https://github.com/BLAST-ImpactX/impactx-benchmarks/blob/main/drift_impactx.py) example, the following timings for `track_particles` on 1, 2, and 4 threads:

`1.893`
`1.017`
`0.5939`

Fix #847

atmyers committed:
Split up particles over multiple tiles for OpenMP

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,25 +141,52 @@ namespace impactx
+    auto n_logical =  numTilesInBox(ba[gid], true, tile_size);
+
+    if (n_logical < nthreads ) {
+        amrex::Print() << "Too few tiles for the number of OpenMP threads. Parallelization will be poor. \n";
```
To avoid stdout flooding, we can use `#include <ablastr/warn_manager/WarnManager.H>`, e.g.,
https://github.com/BLAST-ImpactX/impactx/blob/61b13e453ea5e953b8a5f985539d3e6833cf088f/src/diagnostics/CovarianceMatrixMath.H#L65-L72

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,25 +141,52 @@ namespace impactx
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    {
+        int tid = omp_get_thread_num();
+        int nr = np / nthreads;
+        int nlft = np - nr*nthreads;
+
+        int num_to_add = 0;
+        int my_index =0;
+        if (tid < nlft) { // get nr+1 items
+        my_index = tid * (nr+1);
+        num_to_add = nr+1;
+        } else {         // get nr items
```
Can you please add inline comments on the meaning of the temporary variables (tid, nr, nlft, my_index), maybe spell them out for clarity, and a bit more details on the if/else? :pray:

ax3l reviewed (commented):
Yay, thank you for the fix! :tada:

ax3l commented:
Oh, CI indicates we did not initialize the right number of total particles yet (or forgot them in output)?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -181,25 +208,27 @@ namespace impactx
-            x_arr[old_np+i] = x_ptr[i];
-            y_arr[old_np+i] = y_ptr[i];
-            t_arr[old_np+i] = t_ptr[i];
+            x_arr[old_np+i] = x_ptr[my_index+i];
```
Does `my_index` only need to be added to the index on the LHS? e.g.,
```suggestion
            x_arr[old_np+my_index+i] = x_ptr[i];
```

The tile (`x_ptr`) is shorter now but the `x_arr` is still the one for the whole process?

atmyers committed:
apply suggestions from code review

atmyers committed:
Use better heuristic for setting tile size

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

atmyers commented:
Only the `run_fodo_programmable` fails now. Are we sure that example is thread-safe?

ax3l commented:
Yay! :rocket: :sparkles: 

Yes, I think it is actually not thread safe... Let's set those two controls to `False`:
https://github.com/BLAST-ImpactX/impactx/blob/77980670e525bf07e58fcfb1409feb46c175d3d3/examples/fodo_programmable/run_fodo_programmable.py#L120
https://github.com/BLAST-ImpactX/impactx/blob/77980670e525bf07e58fcfb1409feb46c175d3d3/examples/fodo_programmable/run_fodo_programmable.py#L129

ax3l committed:
FODO Programmable: Not Threadsafe

Needs some further work, e.g., on testing GIL-free Python,
to support threading of Python part.

ax3l committed:
Fix C++ Formatting

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -107,6 +108,49 @@ namespace impactx
+        nthreads = omp_get_max_threads();
+#endif
+
+        const auto& ba = ParticleBoxArray(lid);
+        auto n_logical = numTilesInBox(ba[gid], true, tile_size);
+
+        int ntry = 0;
+        while ((n_logical < nthreads) && (ntry++ < 6)) {
+            int idim = (ntry % 2) + 1;  // alternate between 1 and 2
+            tile_size[idim] /= 2;
+            n_logical = numTilesInBox(ba[gid], true, tile_size);
+        }
```
@atmyers can you please add more inline comments to this block (line 133-141)?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +185,93 @@ namespace impactx
-            ParticleType::NextID(pid+np);
+
+        // split up particles over nthreads tiles
+        AMREX_ALWAYS_ASSERT(numTilesInBox(ParticleBoxArray(lid)[gid], true, tile_size) >= nthreads);
```
@atmyers can you please add a message to this assert?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +185,93 @@ namespace impactx
+
+            // we split up the np particles onto multiple tiles.
+            // some tiles will get nr and some will get nlft.
+            int nr = np / nthreads;
+            int nlft = np - nr*nthreads;
```
@atmyers can you please spell out `nr` and `nlft`?
Do they stand for `n_right` and `n_left`?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +185,93 @@ namespace impactx
+            int num_to_add = 0; /* how many particles this tile will get*/
+            int my_index = 0;
+
+            if (tid < nlft) { // get nr+1 items
+                my_index = tid * (nr + 1);
+                num_to_add = nr + 1;
+            } else {         // get nr items
```
@atmyers can you please add an inline comment here? Who do some get `+1`? Due to rounding into `nthreads` chunks?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +185,93 @@ namespace impactx
+            int nlft = np - nr*nthreads;
+
+            int num_to_add = 0; /* how many particles this tile will get*/
+            int my_index = 0;
```
@atmyers let us pick a more descriptive variable name. Does `my_index` mean `my_thread_index` or `my_tile_index` or something else?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -107,6 +108,49 @@ namespace impactx
+        }
+
+        if (n_logical < nthreads) {
+            amrex::Abort("ImpactParticleContainer::prepare() "
```
@atmyers I added more more concrete guidance to users here, can you please double check it?

In which situations would we not be able to find enough tiles? Is there other guidance we should give or can we make it a `high` warning and "just" be less parallel (i.e., less parallely efficient)?

atmyers replied:
Currently, the code will not just be efficient, it will give incorrect results if the number of tiles according to the `tile_size` is fewer than the number of threads. The issue is that the copyParticles routine in AMReX does not copy the ones that aren't on a valid tile. This resulted in some of the particles not getting written out some examples (since we us copyParticles to a pinned_pc in IO). 

This can be changed in AMReX, I'd suggest changing this to a warning rather than an Abort at that point.

atmyers commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -107,6 +108,49 @@ namespace impactx
+        int ntry = 0;
+        while ((n_logical < nthreads) && (ntry++ < 6)) {
+            int idim = (ntry % 2) + 1;  // alternate between 1 and 2
+            tile_size[idim] /= 2;
```
```suggestion
            tile_size[idim] /= 2;
            AMREX_ALWAYS_ASSERT(tile_size[idim] > 0);
```

atmyers committed:
Update src/particles/ImpactXParticleContainer.cpp

atmyers commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -107,6 +108,50 @@ namespace impactx
+#if defined(AMREX_USE_OMP)
+        nthreads = omp_get_max_threads();
+#endif
+
```
```suggestion
    /**
        When running without space charge and OpenMP parallelization, we need to make sure we have enough tiles on level 0, grid 0 to thread over the available tiles. We try to set the tile_size appropriately here. The tiles start as (long, 8, 8) in (x, y, z). So we try to reduce the tile size in the y and z directions alternately until there are enough tiles for the number of threads. 
    */
```

atmyers committed:
Update src/particles/ImpactXParticleContainer.cpp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

atmyers commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +188,93 @@ namespace impactx
-            ParticleType::NextID(pid+np);
+
+        // split up particles over nthreads tiles
+        AMREX_ALWAYS_ASSERT(numTilesInBox(ParticleBoxArray(lid)[gid], true, tile_size) >= nthreads);
```
```suggestion
        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(numTilesInBox(ParticleBoxArray(lid)[gid], true, tile_size) >= nthreads, "Not enough tiles for the number of OpenMP threads - please try reducing particles.tile_size or increasing the number of cells in the domain.");
```

atmyers committed:
Update src/particles/ImpactXParticleContainer.cpp

atmyers committed:
code review changes

atmyers committed:
forgot to initialize variable

atmyers committed:
add message

atmyers committed:
finish renaming

ax3l reviewed (approved):
Thank you, I understand the logic way better now :pray:

ax3l commented:
@atmyers I see segfaults on Windows right now, which is one of the few runners compiling with `-DImpactX_COMPUTE=NOACC`.

Can you compile in the same mode please and run valgrind on the failing example(s)? I suspect there is an actual bug somewhere.

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +197,96 @@ namespace impactx
+                num_to_add = n_regular;
+            }
+
+            auto& particle_tile = ParticlesAt(lid, gid, tid);
```
@atmyers this line throws if compiled with `-DImpactX_COMPUTE=NOACC`

ax3l replied:
Fixed by defaulting `tid` to `0` not `1`.

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -141,61 +197,96 @@ namespace impactx
-            qm_arr[old_np+i] = qm;
-            w_arr[old_np+i]  = bchchg/ablastr::constant::SI::q_e/np;
-        });
+            int tid = 1;
```
Is the thread index zero-indexed?
```suggestion
            int tid = 0;
```

ax3l replied:
Pushed this change.

ax3l committed:
Fix: `NOACC` with zero-based thread index
